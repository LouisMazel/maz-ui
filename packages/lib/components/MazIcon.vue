<template>
  <!-- eslint-disable vue/no-v-html -->
  <!-- eslint-disable vue/html-self-closing -->
  <svg
    v-if="svgElSource"
    ref="svgElem"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  ></svg>
  <!-- eslint-enable vue/no-v-html -->
  <!-- eslint-enable vue/html-self-closing -->
</template>

<script lang="ts" setup>
  import { nextTick, onMounted, ref, computed, watchEffect, type PropType } from 'vue'

  import { injectStrict } from '../modules/helpers/inject-strict'

  const cache: Record<string, Promise<SVGElement>> = {}
  const svgElSource = ref<SVGElement>()
  const svgElem = ref<SVGElement>()

  const getMazIconPath = () => {
    try {
      return injectStrict<string>('mazIconPath')
    } catch {
      return undefined
    }
  }

  const props = defineProps({
    src: { type: String, default: undefined },
    path: { type: String, default: undefined },
    name: { type: String, default: undefined },
    size: { type: String, default: undefined },
    title: { type: String, default: undefined },
    transformSource: {
      type: Function as PropType<(param: SVGElement) => typeof param>,
      default: (svg: SVGElement) => svg,
    },
  })

  const emits = defineEmits(['loaded', 'unloaded', 'error'])

  const iconPath = computed(() => props.path ?? getMazIconPath())
  const fullSrc = computed(() => {
    if (props.src) {
      return props.src
    } else if (iconPath.value) {
      return `${iconPath.value}/${props.name}.svg`
    } else {
      return `/${props.name}.svg`
    }
  })

  onMounted(() => {
    if (!props.name && !props.src) {
      console.error('[maz-ui](MazIcon) you should provide "name" or "src" as prop')
    }
  })

  const setTitle = (svg: SVGElement, title: string) => {
    const titleTags = svg.querySelectorAll('title')
    if (titleTags.length > 0) {
      // overwrite existing title
      titleTags[0].textContent = title
    } else {
      // create a title element if one doesn't already exist
      const titleEl = document.createElementNS('http://www.w3.org/2000/svg', 'title')
      titleEl.textContent = title
      svg.append(titleEl)
    }
  }

  const filterAttrs = (attrs: Record<string, unknown>) => {
    return Object.keys(attrs).reduce((result, key) => {
      if (attrs[key] !== false && attrs[key] !== null && attrs[key] !== undefined) {
        result[key] = attrs[key]
      }
      return result
    }, {})
  }

  const getSvgAttrs = (svgEl: SVGElement) => {
    // copy attrs
    const svgAttrs = {}
    const attrs = svgEl.attributes
    if (!attrs) {
      return svgAttrs
    }
    for (let i = attrs.length - 1; i >= 0; i--) {
      svgAttrs[attrs[i].name] = attrs[i].value
    }
    return svgAttrs
  }

  const getSvgContent = (svgEl: SVGElement) => {
    let svgElNode = svgEl.cloneNode(true)
    svgElNode = props.transformSource(svgEl)

    if (props.title) {
      setTitle(svgElNode as SVGElement, props.title)
    }

    // copy inner html
    return svgEl.innerHTML
  }

  const getSource = async (src: string) => {
    // fill cache by src with promise
    if (!cache[src]) {
      // download
      cache[src] = download(src)
    }

    // inline svg when cached promise resolves
    try {
      svgElSource.value = await cache[src]
      // wait to render
      await nextTick()
      emits('loaded', svgElem.value)
    } catch (error) {
      if (svgElSource.value) {
        svgElSource.value = undefined
        emits('unloaded')
      }
      // remove cached rejected promise so next image can try load again
      delete cache[src]
      emits('error', error)
    }
  }

  const download = (url: string): Promise<SVGElement> => {
    return new Promise((resolve, reject) => {
      const request = new XMLHttpRequest()
      request.open('GET', url, true)

      request.addEventListener('load', () => {
        if (request.status >= 200 && request.status < 400) {
          try {
            // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
            const parser = new DOMParser()
            const result = parser.parseFromString(request.responseText, 'text/xml')
            let svgEl = result.querySelectorAll('svg')[0] as SVGElement
            if (svgEl) {
              svgEl = props.transformSource(svgEl)
              resolve(svgEl)
            } else {
              reject(new Error('Loaded file is not valid SVG"'))
            }
          } catch (error) {
            reject(error)
          }
        } else {
          reject(new Error('Error loading SVG'))
        }
      })

      request.addEventListener('error', () => reject())
      request.send()
    })
  }

  watchEffect(() => getSource(fullSrc.value))
</script>
