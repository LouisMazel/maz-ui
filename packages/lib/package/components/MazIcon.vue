<template>
  <!-- eslint-disable vue/no-v-html -->
  <svg
    v-if="svgElSource"
    ref="svgElem"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  ></svg>
  <!-- eslint-enable vue/no-v-html -->
</template>

<!-- // inheritAttrs: false -->
<script lang="ts" setup>
  import { nextTick, onMounted, PropType, ref, watch, computed } from 'vue'
  import { injectStrict } from '../utils/injectStrict'

  const cache: Record<string, Promise<SVGElement>> = {}
  const svgElSource = ref<SVGElement>()
  const svgElem = ref<SVGElement>()

  const mazIconPath = injectStrict<string>('mazIconPath')

  const props = defineProps({
    src: { type: String, default: undefined },
    path: { type: String, default: undefined },
    name: { type: String, default: undefined },
    size: { type: String, default: '1.5rem' },
    title: { type: String, default: undefined },
    transformSource: {
      type: Function as PropType<(param: SVGElement) => typeof param>,
      default: (svg: SVGElement) => svg,
    },
    keepDuringLoading: { type: Boolean, default: true },
  })

  const emits = defineEmits(['loaded', 'unloaded', 'error'])

  const iconPath = computed(() => props.path ?? mazIconPath)
  const fullSrc = computed(() =>
    props.src
      ? props.src
      : iconPath.value
      ? `${iconPath.value}/${props.name}.svg`
      : `/${props.name}.svg`,
  )

  onMounted(() => {
    if (!props.name && !props.src) {
      throw new Error('[MazIcon] you should provide "name" or "src" as prop')
    }
    getSource(fullSrc.value)
  })

  const setTitle = (svg: SVGElement, title: string) => {
    const titleTags = svg.getElementsByTagName('title')
    if (titleTags.length) {
      // overwrite existing title
      titleTags[0].textContent = title
    } else {
      // create a title element if one doesn't already exist
      const titleEl = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'title',
      )
      titleEl.textContent = title
      svg.appendChild(titleEl)
    }
  }

  const filterAttrs = (attrs: Record<string, unknown>) => {
    return Object.keys(attrs).reduce((result, key) => {
      if (
        attrs[key] !== false &&
        attrs[key] !== null &&
        attrs[key] !== undefined
      ) {
        result[key] = attrs[key]
      }
      return result
    }, {})
  }

  const getSvgAttrs = (svgEl: SVGElement) => {
    // copy attrs
    const svgAttrs = {}
    const attrs = svgEl.attributes
    if (!attrs) {
      return svgAttrs
    }
    for (let i = attrs.length - 1; i >= 0; i--) {
      svgAttrs[attrs[i].name] = attrs[i].value
    }
    return svgAttrs
  }

  const getSvgContent = (svgEl: SVGElement) => {
    let svgElNode = svgEl.cloneNode(true)
    svgElNode = props.transformSource(svgEl)

    if (props.title) {
      setTitle(svgElNode as SVGElement, props.title)
    }

    // copy inner html
    return svgEl.innerHTML
  }

  const getSource = async (src: string) => {
    // fill cache by src with promise
    if (!cache[src]) {
      // download
      cache[src] = download(src)
    }

    // inline svg when cached promise resolves
    try {
      const svg = await cache[src]
      svgElSource.value = svg
      // wait to render
      await nextTick()
      emits('loaded', svgElem.value)
    } catch (err) {
      if (svgElSource.value) {
        svgElSource.value = undefined
        emits('unloaded')
      }
      // remove cached rejected promise so next image can try load again
      delete cache[src]
      emits('error', err)
    }
  }

  const download = (url: string): Promise<SVGElement> => {
    return new Promise((resolve, reject) => {
      const request = new XMLHttpRequest()
      request.open('GET', url, true)

      request.onload = () => {
        if (request.status >= 200 && request.status < 400) {
          try {
            // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
            const parser = new DOMParser()
            const result = parser.parseFromString(
              request.responseText,
              'text/xml',
            )
            const svgEl = result.getElementsByTagName('svg')[0]
            if (svgEl) {
              // svgEl = this.transformSource(svgEl);
              resolve(svgEl)
            } else {
              reject(new Error('Loaded file is not valid SVG"'))
            }
          } catch (e) {
            reject(e)
          }
        } else {
          reject(new Error('Error loading SVG'))
        }
      }

      request.onerror = reject
      request.send()
    })
  }

  watch(
    () => props.src,
    () => getSource(fullSrc.value),
  )
</script>
