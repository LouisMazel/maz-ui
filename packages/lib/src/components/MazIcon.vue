<script lang="ts" setup>
import { computed, nextTick, onMounted, ref, watchEffect } from 'vue'

import { useInjectStrict } from '../composables/useInjectStrict'

export interface MazIconProps {
  /** The source path of the SVG file - e.g: `/icons/home.svg` */
  src?: string
  /** The path of the folder where the SVG files are stored - e.g: `/icons` */
  path?: string
  /** The name of the SVG file - e.g: `home` */
  name?: string
  /** The size of the SVG file - e.g: `1em` */
  size?: string
  /** The title of the SVG file - e.g: `Home` */
  title?: string
  /** The function to transform the source of the SVG file - e.g: `(svg) => svg` */
  transformSource?: (svg: SVGElement) => SVGElement
}

const {
  src,
  path,
  name,
  size,
  title,
  transformSource = (svg: SVGElement) => svg,
} = defineProps<MazIconProps>()
const emits = defineEmits<{
  /**
   * emitted when SVG file is loaded
   * @property {SVGElement | undefined} svg the svg element loaded
   */
  (event: 'loaded', svg: SVGElement | undefined): void
  /**
   * emitted when SVG file is not loaded
   */
  (event: 'unloaded'): void
  /**
   * emitted when SVG file is not loaded
   * @property {Error} error the error
   */
  (event: 'error', error: Error): void
}>()
const cache: Record<string, Promise<SVGElement>> = {}
const svgElSource = ref<SVGElement>()
const svgElem = ref<SVGElement>()

function getMazIconPath() {
  try {
    return useInjectStrict<string>('mazIconPath')
  }
  catch {
    return undefined
  }
}

const iconPath = computed(() => path ?? getMazIconPath())
const fullSrc = computed(() => {
  if (src) {
    return src
  }
  else if (iconPath.value) {
    return `${iconPath.value}/${name}.svg`
  }
  else {
    return `/${name}.svg`
  }
})

onMounted(() => {
  if (!name && !src) {
    console.error('[maz-ui](MazIcon) you should provide "name" or "src" as prop')
  }
})

function setTitle(svg: SVGElement, title: string) {
  const titleTags = svg.querySelectorAll('title')
  if (titleTags.length > 0) {
    // overwrite existing title
    titleTags[0].textContent = title
  }
  else {
    // create a title element if one doesn't already exist
    const titleEl = document.createElementNS('http://www.w3.org/2000/svg', 'title')
    titleEl.textContent = title
    svg.append(titleEl)
  }
}

function filterAttrs(attrs: Record<string, unknown>) {
  return Object.keys(attrs).reduce((result, key) => {
    if (attrs[key] !== false && attrs[key] !== null && attrs[key] !== undefined) {
      result[key] = attrs[key]
    }
    return result
  }, {} as Record<string, unknown>)
}

function getSvgAttrs(svgEl: SVGElement) {
  // copy attrs
  const svgAttrs: Record<string, string> = {}
  const attrs = svgEl.attributes
  if (!attrs) {
    return svgAttrs
  }
  for (let i = attrs.length - 1; i >= 0; i--) {
    svgAttrs[attrs[i].name] = attrs[i].value
  }
  return svgAttrs
}

function getSvgContent(svgEl: SVGElement) {
  svgEl.cloneNode(true)
  const svgElNode = transformSource(svgEl)

  if (title) {
    setTitle(svgElNode as SVGElement, title)
  }

  // copy inner html
  return svgEl.innerHTML
}

async function getSource(src: string) {
  // fill cache by src with promise
  if (!cache[src]) {
    // download
    cache[src] = download(src)
  }

  // inline svg when cached promise resolves
  try {
    svgElSource.value = await cache[src]
    // wait to render
    await nextTick()
    emits('loaded', svgElem.value)
  }
  catch (error) {
    if (svgElSource.value) {
      svgElSource.value = undefined
      emits('unloaded')
    }
    // remove cached rejected promise so next image can try load again
    delete cache[src]
    emits('error', error as Error)
  }
}

function download(url: string): Promise<SVGElement> {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest()
    request.open('GET', url, true)

    request.addEventListener('load', () => {
      if (request.status >= 200 && request.status < 400) {
        try {
          // Setup a parser to convert the response to text/xml in order for it to be manipulated and changed
          const parser = new DOMParser()
          const result = parser.parseFromString(request.responseText, 'text/xml')
          let svgEl = result.querySelectorAll('svg')[0] as SVGElement
          if (svgEl) {
            svgEl = transformSource(svgEl)
            resolve(svgEl)
          }
          else {
            reject(new Error('Loaded file is not valid SVG"'))
          }
        }
        catch (error) {
          reject(error)
        }
      }
      else {
        reject(new Error('Error loading SVG'))
      }
    })

    request.addEventListener('error', error => reject(error))
    request.send()
  })
}

watchEffect(() => getSource(fullSrc.value))
</script>

<template>
  <svg
    v-if="svgElSource"
    ref="svgElem"
    class="m-icon m-reset-css"
    width="1em"
    height="1em"
    v-bind="{
      ...getSvgAttrs(svgElSource),
      ...filterAttrs($attrs),
    }"
    :style="`font-size: ${size}`"
    v-html="getSvgContent(svgElSource)"
  />
</template>

<style lang="postcss" scoped>
.m-icon {
  width: 1em !important;
  height: 1em !important;
}
</style>
