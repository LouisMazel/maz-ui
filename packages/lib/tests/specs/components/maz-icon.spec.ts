import { shallowMount } from '@vue/test-utils'
import MazIcon from '@components/MazIcon.vue'

describe('MazIcon', () => {
  it('renders correctly with custom name prop', async () => {
    const wrapper = shallowMount(MazIcon, {
      propsData: {
        name: 'test-icon',
      },
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.fullSrc).toBe('/test-icon.svg')
  })

  it('renders the icon with the specified name and size', () => {
    const wrapper = shallowMount(MazIcon, {
      propsData: {
        name: 'icon-name',
        size: '24px',
      },
    })
    expect(wrapper.find('svg').attributes().width).toBe('24px')
    expect(wrapper.find('svg').attributes().height).toBe('24px')
    expect(wrapper.find('svg').html()).toContain('icon-name')
  })

  it('renders the icon with the specified title', () => {
    const wrapper = shallowMount(MazIcon, {
      propsData: {
        name: 'icon-name',
        title: 'My icon',
      },
    })
    expect(wrapper.find('svg title').text()).toBe('My icon')
  })

  it('throws an error if no name or src prop is provided', () => {
    expect(() => {
      shallowMount(MazIcon)
    }).toThrowError(
      '[maz-ui](MazIcon) you should provide "name" or "src" as prop',
    )
  })

  it('renders the icon using the specified source URL', () => {
    const wrapper = shallowMount(MazIcon, {
      propsData: {
        src: 'http://example.com/icon.svg',
      },
    })
    expect(wrapper.find('svg').attributes().width).toBe('1em')
    expect(wrapper.find('svg').attributes().height).toBe('1em')
    expect(wrapper.find('svg').html()).toContain('http://example.com/icon.svg')
  })
})
