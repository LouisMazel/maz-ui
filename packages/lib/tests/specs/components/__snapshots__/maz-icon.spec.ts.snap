// Vitest Snapshot v1

exports[`MazIcon > renders correctly with custom name prop 1`] = `
<div
  data-v-app=""
>
  
  <!-- eslint-disable -->
  <!--v-if-->
  <!-- eslint-enable -->
  
</div>
`;
